package fetch.cfg.tasks;

message AutoTile
{
  optional double z_um                = 1 [default = 100.0];
  optional double maxz_mm             = 2 [default = 10.0];
  optional uint32 timeout_ms          = 3 [default = 5000];
  optional uint32 ichan               = 4 [default = 0];
  optional double intensity_threshold = 5 [default = 1000];
  optional double area_threshold      = 6 [default = 0.1];
  
  optional int32  search_radius       = 7 [default = 3];    // search radius in tiles
  optional bool   use_adaptive_tiling = 8 [default = false];
  optional bool   use_explore         = 9 [default = true];

  optional bool   schedule_stop_after_nth_cut = 10 [default = false];
  optional uint32 nth_cut_to_stop_after		  = 11 [default = 1];
  optional uint32 cut_count_since_scheduled_stop = 12 [default = 0];
}

message TimeSeries
{ optional double interval_ms         = 1 [default = 0.0];
}

message SurfaceFind
{
  optional double backup_frac  = 1 [default = 0.1]; // step stage out by stack depth * backup_frac
  optional double max_backup_um =2 [default = 500]; // limit the total search space
  optional double max_raise_um  =3 [default = 50]; // limit the amount the stage can be moved up during surface find

  optional double min_um       = 4 [default = 0];   // 
  optional double max_um       = 5 [default = 100]; // scan a stack from min to max by dz, inclusive
  optional double dz_um        = 6 [default = 5];   // 
  
  optional uint32 ichan               = 7 [default = 0];    //
  optional double intensity_threshold = 8 [default = 1000]; // Classify as in or out
  optional double area_threshold      = 9 [default = 0.1];  //

  optional double offset_um           =10 [default = 0];    // adjust tile offset to image this many microns below the cut
  //optional double response            = 11[default = 2.0]; // correct stage z by response * measured error
  //optional double response_limit_frac = 12[default = 0.1]; // response is limited to response_limit_frac * stack depth
}

message AdaptiveTiling {
  optional int32 every   = 1 [default=100]; // check every "every" tiles
  optional int32 mindist = 2 [default=3];   // avoid tiles where the edge of the active area is within mindist tiles
  optional double initial_distance_from_surface_um = 3 [default=-20]; //DGA: the distance from the surface that the stage should move to before moving to the proper surface height
}